src/Microsoft.DotNet.Tools.Run/RunCommand.cs

```diff
@@ -0,0 +1,154 @@
+            }
+            else
+            {
+                var fx = NuGetFramework.Parse(Framework);
+                _context = contexts.FirstOrDefault(c => c.TargetFramework.Equals(fx));
+            }
+
+            if (Args == null)
+            {
+                _args = new List<string>();
+            }
+            else
+            {
+                _args = new List<string>(Args);
+            }
+        }
+
+        private int RunExecutable()
+        {
+            // Create a temporary directory
+            var tempDir = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString("N"));
+
+            // Compile to that directory
+            var result = Command.Create($"dotnet-compile", $"--output \"{tempDir}\" --temp-output \"{tempDir}\" --framework \"{_context.TargetFramework}\" --configuration \"{Configuration}\" {_context.ProjectFile.ProjectDirectory}")
+                .ForwardStdOut(onlyIfVerbose: true)
+                .ForwardStdErr()
+                .Execute();
+
+            if (result.ExitCode != 0)
+            {
+                return result.ExitCode;
+            }
+
+            // Now launch the output and give it the results
+            var outputName = Path.Combine(tempDir, _context.ProjectFile.Name + Constants.ExeSuffix);
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                if (_context.TargetFramework.IsDesktop())
+                {
+                    // Run mono if we're running a desktop target on non windows
+                    _args.Insert(0, outputName + ".exe");
+
+                    if (string.Equals(Configuration, "Debug", StringComparison.OrdinalIgnoreCase))
+                    {
+                        // If we're compiling for the debug configuration then add the --debug flag
+                        // other options may be passed using the MONO_OPTIONS env var
+                        _args.Insert(0, "--debug");
+                    }
+
+                    outputName = "mono";
+                }
+            }
+
+            // Locate the runtime
+            string runtime = Environment.GetEnvironmentVariable("DOTNET_HOME");
+            if (string.IsNullOrEmpty(runtime))
+            {
+                // Use the runtime deployed with the tools, if present
+                var candidate = Path.Combine(AppContext.BaseDirectory, "..", "runtime");
+                if (File.Exists(Path.Combine(candidate, Constants.LibCoreClrName)))
+                {
+                    runtime = Path.GetFullPath(candidate);
+                }
+            }
+
+            result = Command.Create(outputName, string.Join(" ", _args))
+                .ForwardStdOut()
+                .ForwardStdErr()
+                .EnvironmentVariable("DOTNET_HOME", runtime)
+                .Execute();
+
+            // Clean up
+            if (!PreserveTemporary)
+            {
+                Directory.Delete(tempDir, recursive: true);
```

Comment:
I'll address this separately (either will create another PR just after this or create an issue)